course-content/31_exploring_data_structures.qmd

```diff
@@ -0,0 +1,293 @@
+
+:::
+
+
+
+:::{.callout-note icon="false" collapse="true"}
+
+### Solution to Challenge 2
+
+First you would like to create a new object to hold this narrowed down version of the dataset, then just use some method like `.print()` or `.head()` to see some of the resulting data. 
+
+```python
+challenge2_df = df.query('country.str.startswith("Canada") & year == 1972')
+print(challenge2_df)
+```
+
+:::
+
+:::{.callout-important}
+There is something else you should know about row selection. It is VERY common within python/pandas to see the `.loc()` and `.iloc()` codes to identify rows by index location.   The index starts at 0 for the first row (yes, "0", and not "1")!  And you can also use this method You may not ever need to use it, but it is good to know what it means when you come across it. 
+:::
+
+```{python}
+x_df = df.loc[0:1]
+print(x_df)
+```
+
+## 3.1.4. Sorting rows
+
+So now you've mastered how to select rows and columns you want, congratulations!
+
+Instead of hunting and pecking for insights, one way to quickly make some sense of the data is to sort it - something you probably do in Excel all the time. 
+
+Above, we have created a dataframe with just the one country in it and for all years above 1972. So now let's sort the data by year, and then see what can spot any insights. To do this, we use the `.sort_values()` method.
+
+Let's go fetch and look at the filtered_df dataframe from above, and use this method to sort by year, going from more recent to less recent:
+
+```{python}
+filtered_df.sort_values('lifeExp', ascending=False)
+#
+```
+
+Here we can read a few insights out of the data.
+
+## 3.1.5. Putting multiple methods together (chaining)
+
+You can also put multiple methods in the same line of code, as long as you separate each of the elements. In pandas, this is called "method chaining",  as you are kind of creating a chain of actions to take place. 
+
+```{python}
+#| scrolled: true
+filtered_df = df.query('country.str.startswith("Ca") & year == 2007').sort_values('year', ascending = False)
+filtered_df.head(3)
+```
+
+:::{.callout-warning icon="false"}
+
+### Challenge 3
+
+
+
+Your director has come back to you and wondered aloud whether the life expectancy of people changed during the 1970s in Cambodia. Use what you know about selecting rows and sorting data to get the data you need to answer the question. 
+
+:::
+
+
+
+:::{.callout-note icon="false" collapse="true"}
+
+### Solution to Challenge 3
+
+Using the query() and sort.values() methods and by chaining them together. Looking at the data we see that in the 1970s there was a sharp decline in life expectancy in Cambodia. We also see, thankfully, that is has recovered strongly since then. 
+
+```python
+challenge3_df = df.query('country.str.startswith("Cambodia") & year < 2008').sort_values('year', ascending = False)
+print(challenge3_df)
+```
+
+:::
+
+## 3.1.6. Creating new columns of data
+
```

Comment:
Good idea.

Reply:
I might just reference them in speaking and live code...

Reply:
I think if it's going in the live code it would be a useful reference to include in here. Is assign your go-to-method as well? I'm much in the camp of the other 3 types, but should stick to what you're most comfortable with.

Reply:
Good point. I'll incorporate them on the page and they'll be there for easier reference and speaking to